### PR TITLE
[web] Rich text painting on bitmap canvas

### PR DESCRIPTION
**Pull request by mdebbar**
## Description

This PR implements the rendering part for rich paragraphs.

Goldens PR: https://github.com/flutter/goldens/pull/132

### CODE REVIEW

File: lib/web_ui/lib/src/engine/text/layout_service.dart [hunk 1]
```diff
@@ -312,6 +312,10 @@ class RangeBox {
     return startIndex < this.end.index && this.start.index < endIndex;
   }
 
+  ui.TextBox toTextBox(EngineLineMetrics line) {
```
Comment (ferhatb) [inline lib/web_ui/lib/src/engine/text/layout_service.dart:315]: document

File: lib/web_ui/lib/src/engine/text/paint_service.dart [hunk 2]
```diff
@@ -0,0 +1,72 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+// @dart = 2.12
+part of engine;
+
+class TextPaintService {
```
Comment (ferhatb) [inline lib/web_ui/lib/src/engine/text/paint_service.dart:8]: Add comment:
Loop through the spans and for each box in the span:
    // 1. Paint the background rect.
    // 2. Paint the text shadows?
    // 3. Paint the text.

**Review by ferhatb — APPROVED** [review]

File: lib/web_ui/lib/src/engine/text/paint_service.dart (hunk 2 above)
Comment (mdebbar) [reply lib/web_ui/lib/src/engine/text/paint_service.dart:9]: This comment was more of a pseudo-implementation with the TODO. The current implementation is different from how I imagined it at the time of this comment. Right now the boxes are part of the "line" not the "span", for example. This made more sense during the implementation of gesture recognition and painting.

Added multiple comments to this file.

File: lib/web_ui/lib/src/engine/text/layout_service.dart (hunk 1 above)
Comment (mdebbar) [reply lib/web_ui/lib/src/engine/text/layout_service.dart:319]: Done.

**Pull request merged**